### PR TITLE
Pippa/query pe

### PR DESCRIPTION
**Pull request by pippaduckett1**

### CODE REVIEW

File: src/hepattn/models/maskformer.py [hunk 1]
```diff
@@ -126,11 +213,30 @@ def forward(self, inputs: dict[str, Tensor]) -> dict[str, Tensor]:
         x["query_embed"] = self.query_initial.expand(batch_size, -1, -1)
         x["query_valid"] = torch.full((batch_size, self.num_queries), True, device=x["query_embed"].device)
 
+        # Add positional encoding to the queries
```
Comment (samvanstroud) [inline src/hepattn/models/maskformer.py:216]: I'd extract this to a function `add_query_posenc` and call it right before (or inside) the decoder loop

**Comment by samvanstroud** [comment]
Thanks @pippaduckett1! Had a quick first look, I was taking a look as the PR updated so not sure if anything will be useful

I'd move the attention mask logging fixes to another PR if possible :)

File: src/hepattn/models/posenc.py [hunk 2]
```diff
@@ -135,3 +135,71 @@ def forward(self, xs: dict[str, Tensor]) -> Tensor:
         xs = 2 * self.pi * xs
         xs @= self.B
         return torch.cat([torch.sin(xs), torch.cos(xs)], dim=-1)
+
+
+class QueryPositionEncoder(nn.Module):
```
[the remark this answers is not in the document]
Comment (pippaduckett1) [reply src/hepattn/models/posenc.py:140]: Definitely agree - this is legacy from when I was trying something different - will remove :)

File: src/hepattn/models/maskformer.py [hunk 3]
```diff
@@ -114,6 +198,9 @@ def forward(self, inputs: dict[str, Tensor]) -> dict[str, Tensor]:
 
         # Pass merged input hits through the encoder
         if self.encoder is not None:
+            if self.preserve_original_keys:
+                x["key_embed_original"] = x["key_embed"].clone()
+
             # Note that a padded feature is a feature that is not valid!
```
[the remark this answers is not in the document]
Comment (pippaduckett1) [reply src/hepattn/models/maskformer.py:204]: yes - this is from Segment Anything
Screenshot 2025-07-21 at 12 59 32

**Comment by samvanstroud** [comment]
@pippaduckett1 are we closing this in favour of #70

File: src/hepattn/models/maskformer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (samvanstroud) [reply src/hepattn/models/maskformer.py:204]: I'm not sure the prompt tokens are analagous to our kv tokens here. Our kv tokens are more like the image pixels for SAM, and SAM' prompt is something we don't really have in our setup
